File: src/main/java/io/realworld/api/ArticlesApi.java
package io.realworld.api;

import io.realworld.model.CreateArticle201Response;
import io.realworld.model.CreateArticleRequest;
import io.realworld.model.GetArticlesFeed200Response;
import io.realworld.model.UpdateArticleRequest;

import io.vertx.core.Future;

public interface ArticlesApi  {
    Future<ApiResponse<CreateArticle201Response>> createArticle(CreateArticleRequest article);
    Future<ApiResponse<Void>> deleteArticle(String slug);
    Future<ApiResponse<CreateArticle201Response>> getArticle(String slug);
    Future<ApiResponse<GetArticlesFeed200Response>> getArticles(String tag, String author, String favorited, Integer offset, Integer limit);
    Future<ApiResponse<GetArticlesFeed200Response>> getArticlesFeed(Integer offset, Integer limit);
    Future<ApiResponse<CreateArticle201Response>> updateArticle(String slug, UpdateArticleRequest article);
}

File: src/main/kotlin/io/realworld/api/ArticlesApi.kt
package io.realworld.api

import io.realworld.model.request.CreateArticleRequest
import io.realworld.model.request.UpdateArticleRequest
import io.realworld.model.response.ApiResponse
import io.realworld.model.response.CreateArticle201Response
import io.realworld.model.response.GetArticlesFeed200Response
import io.vertx.core.Future

interface ArticlesApi {
  fun createArticle(article: CreateArticleRequest): Future<ApiResponse<CreateArticle201Response>>
  fun deleteArticle(slug: String): Future<ApiResponse<Void>>
  fun getArticle(slug: String): Future<ApiResponse<CreateArticle201Response>>
  fun getArticles(
    tag: String?,
    author: String?,
    favorited: String?,
    offset: Int,
    limit: Int
  ): Future<ApiResponse<GetArticlesFeed200Response>>

  fun getArticlesFeed(offset: Int, limit: Int): Future<ApiResponse<GetArticlesFeed200Response>>
  fun updateArticle(slug: String, article: UpdateArticleRequest): Future<ApiResponse<CreateArticle201Response>>
}
